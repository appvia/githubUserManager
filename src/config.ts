export const config = {
  get addUsers(): boolean {
    return process.env.ADD_USERS?.toLowerCase() === 'true'
  },
  get removeUsers(): boolean {
    return process.env.REMOVE_USERS?.toLowerCase() === 'true'
  },
  get exitCodeOnMissmatch(): number {
    return parseInt(process.env.EXIT_CODE_ON_MISMATCH ?? '0') || 0
  },
  get ignoredUsers(): string[] {
    return process.env.IGNORED_USERS?.toLowerCase().split(',') ?? []
  },
  get githubPrivateKey(): string {
    return Buffer.from(process.env.GITHUB_PRIVATE_KEY, 'base64').toString('utf-8')
  },
  get githubAppID(): number {
    return parseInt(process.env.GITHUB_APP_ID)
  },
  get githubInstallationID(): number {
    return parseInt(process.env.GITHUB_INSTALLATION_ID)
  },
  get githubOrg(): string {
    return process.env.GITHUB_ORG ?? ''
  },
  get googleCredentials(): googleCredentials {
    return JSON.parse(Buffer.from(process.env.GOOGLE_CREDENTIALS, 'base64').toString('utf-8'))
  },
  get googleEmailAddress(): string {
    return process.env.GOOGLE_EMAIL_ADDRESS ?? ''
  },
}

export interface googleCredentials {
  type: string
  project_id: string
  private_key_id: string
  private_key: string
  client_email: string
  client_id: string
  auth_uri: string
  token_uri: string
  auth_provider_x509_cert_url: string
  client_x509_cert_url: string
}
