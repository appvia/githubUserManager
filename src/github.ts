import { createAppAuth } from '@octokit/auth-app'
import { Octokit } from '@octokit/rest'
import * as mod from './github'
import { config } from './config'
import { GetResponseTypeFromEndpointMethod } from '@octokit/types'

export function getAuthenticatedOctokit(): Octokit {
  return new Octokit({
    authStrategy: createAppAuth,
    auth: {
      appId: config.githubAppID,
      privateKey: config.githubPrivateKey,
      installationId: config.githubInstallationID,
    },
  })
}

export async function getGithubUsersFromGithub(): Promise<Set<string>> {
  const octokit = mod.getAuthenticatedOctokit()
  const members = await octokit.paginate(octokit.orgs.listMembers, {
    org: config.githubOrg,
  })

  const pendingInvites = await octokit.paginate(octokit.orgs.listPendingInvitations, {
    org: config.githubOrg,
  })
  const pendingGithubAccounts = formatUserList(pendingInvites)

  const githubAccounts = formatUserList(members)

  if (pendingGithubAccounts.size > 0)
    console.log(`Outstanding GitHub invites for ${Array.from(pendingGithubAccounts).join(', ')}`)

  return new Set([...githubAccounts, ...pendingGithubAccounts])
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export function formatUserList(users): Set<string> {
  return new Set(
    users
      .map((user) => user.login?.toLowerCase())
      .flat()
      .filter(Boolean),
  )
}

export async function getUserIdFromUsername(username: string): Promise<number> {
  const octokit = mod.getAuthenticatedOctokit()
  console.log(`Looking up user ${username}`)
  let user
  try {
    user = await octokit.users.getByUsername({ username })
  } catch (error) {
    throw `Unable to find user id for ${username}`
  }
  console.log(`User ${username} userid: ${user.data.id}`)
  return user.data.id
}

export async function addUsersToGitHubOrg(users: Set<string>): Promise<void> {
  for (const user of users) {
    await mod.addUserToGitHubOrg(user)
  }
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export async function addUserToGitHubOrg(
  user: string,
): Promise<GetResponseTypeFromEndpointMethod<typeof octokit.orgs.createInvitation> | boolean> {
  const octokit = mod.getAuthenticatedOctokit()
  if (config.ignoredUsers.includes(user.toLowerCase())) {
    console.log(`Ignoring add for ${user}`)
    return false
  }
  const userId = await mod.getUserIdFromUsername(user)
  console.log(`Inviting ${user} (${userId} to ${config.githubOrg})`)
  return await octokit.orgs.createInvitation({
    org: config.githubOrg,
    invitee_id: userId,
  })
}

export async function removeUsersFromGitHubOrg(users: Set<string>): Promise<void> {
  for (const user of users) {
    await mod.removeUserFromGitHubOrg(user)
  }
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export async function removeUserFromGitHubOrg(
  user: string,
): Promise<GetResponseTypeFromEndpointMethod<typeof octokit.orgs.removeMembershipForUser> | boolean> {
  const octokit = mod.getAuthenticatedOctokit()
  if (config.ignoredUsers.includes(user.toLowerCase())) {
    console.log(`Ignoring remove for ${user}`)
    return false
  }
  console.log(`Removing user/invitation ${user} from ${config.githubOrg}`)
  return octokit.orgs.removeMembershipForUser({
    org: config.githubOrg,
    username: user,
  })
}
