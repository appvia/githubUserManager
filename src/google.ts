import { google } from 'googleapis'
import * as mod from './google'
import { config } from './config'

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export async function googleAuth() {
  const privatekey = config.googleCredentials
  const jwtClient = new google.auth.JWT(
    privatekey.client_email,
    null,
    privatekey.private_key,
    ['https://www.googleapis.com/auth/admin.directory.user.readonly'],
    config.googleEmailAddress,
  )
  await jwtClient.authorize()
  return jwtClient
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export async function getAdminService() {
  return google.admin({
    version: 'directory_v1',
    auth: await googleAuth(),
  })
}

export async function getGithubUsersFromGoogle(): Promise<Set<string>> {
  const service = await mod.getAdminService()

  const userList = await service.users.list({
    customer: 'my_customer',
    maxResults: 250,
    projection: 'custom',
    fields: 'users(customSchemas/Accounts/github(value))',
    customFieldMask: 'Accounts',
  })

  const githubAccounts = mod.formatUserList(userList.data.users)
  return githubAccounts
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export function formatUserList(users): Set<string> {
  return new Set(
    users
      .map((user) => user.customSchemas?.Accounts?.github?.map((account) => account.value?.toLowerCase()))
      .flat()
      .filter(Boolean),
  )
}
