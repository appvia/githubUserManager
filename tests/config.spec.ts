import * as mod from '../src/config'

describe('addUsers', () => {
  beforeEach(() => {
    delete process.env.ADD_USERS
  })
  it('no value', () => {
    expect(mod.config.addUsers).toStrictEqual(false)
  })
  it('invalid value', () => {
    process.env.ADD_USERS = 'foobar'
    expect(mod.config.addUsers).toStrictEqual(false)
  })
  it('false value', () => {
    process.env.ADD_USERS = 'false'
    expect(mod.config.addUsers).toStrictEqual(false)
  })
  it('true value', () => {
    process.env.ADD_USERS = 'true'
    expect(mod.config.addUsers).toStrictEqual(true)
  })
  it('true value mixed case', () => {
    process.env.ADD_USERS = 'tRuE'
    expect(mod.config.addUsers).toStrictEqual(true)
  })
})

describe('removeUsers', () => {
  beforeEach(() => {
    delete process.env.REMOVE_USERS
  })
  it('no value', () => {
    expect(mod.config.removeUsers).toStrictEqual(false)
  })
  it('invalid value', () => {
    process.env.REMOVE_USERS = 'foobar'
    expect(mod.config.removeUsers).toStrictEqual(false)
  })
  it('false value', () => {
    process.env.REMOVE_USERS = 'false'
    expect(mod.config.removeUsers).toStrictEqual(false)
  })
  it('true value', () => {
    process.env.REMOVE_USERS = 'true'
    expect(mod.config.removeUsers).toStrictEqual(true)
  })
  it('true value mixed case', () => {
    process.env.REMOVE_USERS = 'tRuE'
    expect(mod.config.removeUsers).toStrictEqual(true)
  })
})

describe('ignoredUsers', () => {
  it('getIgnoredUsers empty', () => {
    delete process.env.IGNORED_USERS
    expect(mod.config.ignoredUsers).toMatchSnapshot()
  })

  it('getIgnoredUsers single', () => {
    process.env.IGNORED_USERS = 'user1'
    expect(mod.config.ignoredUsers).toMatchSnapshot()
  })
  it('getIgnoredUsers many', () => {
    process.env.IGNORED_USERS = 'user1,user2,user3,USER4'
    expect(mod.config.ignoredUsers).toMatchSnapshot()
  })
})

describe('githubPrivateKey', () => {
  it('should base 64 decode', () => {
    process.env.GITHUB_PRIVATE_KEY = Buffer.from('helloworld').toString('base64')
    expect(mod.config.githubPrivateKey).toStrictEqual('helloworld')
  })
})

describe('googleCredentials', () => {
  it('should base 64 decode and json parse', () => {
    process.env.GOOGLE_CREDENTIALS = Buffer.from(JSON.stringify({ hello: 'world' })).toString('base64')
    expect(mod.config.googleCredentials).toStrictEqual({ hello: 'world' })
  })
})

describe('exitCodeOnMissmatch', () => {
  beforeEach(() => {
    delete process.env.EXIT_CODE_ON_MISMATCH
  })
  it('no value', () => {
    expect(mod.config.exitCodeOnMissmatch).toStrictEqual(0)
  })
  it('invalid value', () => {
    process.env.EXIT_CODE_ON_MISMATCH = 'hello'
    expect(mod.config.exitCodeOnMissmatch).toStrictEqual(0)
  })
  it('valid value', () => {
    process.env.EXIT_CODE_ON_MISMATCH = '123'
    expect(mod.config.exitCodeOnMissmatch).toStrictEqual(123)
  })
})

describe('githubAppID', () => {
  beforeEach(() => {
    delete process.env.GITHUB_APP_ID
  })
  it('no value', () => {
    expect(mod.config.githubAppID).toStrictEqual(NaN)
  })
  it('invalid value', () => {
    process.env.GITHUB_APP_ID = 'hello'
    expect(mod.config.githubAppID).toStrictEqual(NaN)
  })
  it('valid value', () => {
    process.env.GITHUB_APP_ID = '123'
    expect(mod.config.githubAppID).toStrictEqual(123)
  })
})

describe('githubInstallationID', () => {
  beforeEach(() => {
    delete process.env.GITHUB_INSTALLATION_ID
  })
  it('no value', () => {
    expect(mod.config.githubInstallationID).toStrictEqual(NaN)
  })
  it('invalid value', () => {
    process.env.GITHUB_INSTALLATION_ID = 'hello'
    expect(mod.config.githubInstallationID).toStrictEqual(NaN)
  })
  it('valid value', () => {
    process.env.GITHUB_INSTALLATION_ID = '123'
    expect(mod.config.githubInstallationID).toStrictEqual(123)
  })
})

describe('githubOrg', () => {
  beforeEach(() => {
    delete process.env.GITHUB_ORG
  })
  it('no value', () => {
    expect(mod.config.githubOrg).toStrictEqual('')
  })
  it('valid value', () => {
    process.env.GITHUB_ORG = 'hello'
    expect(mod.config.githubOrg).toStrictEqual('hello')
  })
})

describe('googleEmailAddress', () => {
  beforeEach(() => {
    delete process.env.GOOGLE_EMAIL_ADDRESS
  })
  it('no value', () => {
    expect(mod.config.googleEmailAddress).toStrictEqual('')
  })
  it('valid value', () => {
    process.env.GOOGLE_EMAIL_ADDRESS = 'hello'
    expect(mod.config.googleEmailAddress).toStrictEqual('hello')
  })
})
