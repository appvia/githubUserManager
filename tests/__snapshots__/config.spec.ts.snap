// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ignoredUsers getIgnoredUsers empty 1`] = `Array []`;

exports[`ignoredUsers getIgnoredUsers many 1`] = `
Array [
  "user1",
  "user2",
  "user3",
  "user4",
]
`;

exports[`ignoredUsers getIgnoredUsers single 1`] = `
Array [
  "user1",
]
`;
